<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
import piplite
await piplite.install(['folium'])
await piplite.install(['pandas'])

In [2]:
import folium
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module\_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [6]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#0000FF', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=4)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label

for index,row in launch_sites_df.iterrows():

    circle = folium.Circle([row["Lat"],row["Long"]], radius=1000, color='#d35400', fill=True).add_child(folium.Popup(row["Launch Site"]))
   
    marker = folium.map.Marker(
        [row["Lat"],row["Long"]],
        # Create an icon as a text label
        icon=DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % row["Launch Site"],
            )
        )
    site_map.add_child(circle)
    site_map.add_child(marker)

equator_line = folium.PolyLine(locations=[[-5, -180], [5, 180]], color='blue', weight=2.5, opacity=1)
site_map.add_child(equator_line)

equator_label = folium.Marker(
    location=[-5, -120],
    icon=folium.DivIcon(html='<div style="font-size: 12; color:blue;"><b>Equator Line</b></div>'),
    tooltip="Equator Line"
)
site_map.add_child(equator_label)

site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


##### Explanation:

The Florida launch sites (CCAFS LC-40, CCAFS SLC-40, and KSC LC-39A) are closer to the Equator than VAFB SLC-4E, which is located in California and farther away.

All launch sites are in proximity to a coast, but the Florida launch sites are on the east coast, and VAFB SLC-4E is on the west coast.

Launching near the Equator maximizes the Earth's rotational speed, providing a natural boost for rockets. This efficiency is crucial for payload capacity, cost-effectiveness, and mission success, making it an important consideration for space agencies, satellite operators, and commercial clientt:

Coastal launch sites offer overwater trajectories, enhancing safety by minimizing populated areas overflown during launches. Additionally, access to the coast facilitates transportation of rocket components and equipment, influencing logistics and operational efficiency for stakeholders involved in space miss






# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [9]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [10]:
marker_cluster = MarkerCluster()


*TODO:* Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [11]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red

def marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'
    
spacex_df['marker_color'] = spacex_df['class'].apply(marker_color)
spacex_df.tail(10)

Launch Site        Lat       Long  class marker_color
46    KSC LC-39A  28.573255 -80.646895      1        green
47    KSC LC-39A  28.573255 -80.646895      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
49  CCAFS SLC-40  28.563197 -80.576820      1        green
50  CCAFS SLC-40  28.563197 -80.576820      1        green
51  CCAFS SLC-40  28.563197 -80.576820      0          red
52  CCAFS SLC-40  28.563197 -80.576820      0          red
53  CCAFS SLC-40  28.563197 -80.576820      0          red
54  CCAFS SLC-40  28.563197 -80.576820      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [12]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    # Create a Marker object with its coordinate
    marker = folium.Marker(
        location=[record['Lat'], record['Long']],
        icon=folium.Icon(color='white', icon_color=record['marker_color']),
        popup=f"Launch Site: {record['Launch Site']}, Outcome: {'Success' if record['class'] == 1 else 'Failure'}"
    )
    # Add the marker to the MarkerCluster
    marker_cluster.add_child(marker)

# Add marker_cluster to the current site_map
site_map.add_child(marker_cluster)

# Display the map
site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [13]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [14]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [15]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
launch_site_lat = 28.563197
launch_site_lon = -80.57682
coastline_lat = 28.56349
coastline_lon = -80.56802
distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
distance_coastline

0.8603066326319084

In [16]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
coastline_marker = folium.Marker(
    location=[coastline_lat, coastline_lon],
    icon=folium.DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:.2f} KM".format(distance_coastline),
    ),
    popup="Closest Coastline"
)
site_map.add_child(coastline_marker)

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [17]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
lines=folium.PolyLine(locations=[[launch_site_lat, launch_site_lon], [coastline_lat, coastline_lon]], weight=1)
site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [ ]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site


In [18]:
def add_marker_and_line(map, site, location, lat1, lon1, lat2, lon2, color):
    distance = calculate_distance(lat1, lon1, lat2, lon2)
    marker = folium.Marker(
        location=[lat2, lon2],
        icon=folium.DivIcon(
            icon_size=(20, 20),
            icon_anchor=(0, 0),
            html='<div style="font-size: 12; color:%s;"><b>%s</b></div>' % (color, "Distance from {} to {} = {:.2f} KM".format(site, location, distance)),
        ),
        popup="{} to {}".format(site, location)
    )
    map.add_child(marker)
    lines = folium.PolyLine(locations=[[lat1, lon1], [lat2, lon2]], weight=1, color=color)
    map.add_child(lines)

# Define the coordinates for each site
sites = {
    'CCAFS LC-40': {
        'launch_site': [28.56230197, -80.57735648],
        'highway': [28.56238, -80.57065],
        'coastline': [28.56253, -80.56786],
        'railroad': [28.56262, -80.5871],
        'city': [28.48921, -80.56274]
    },
    'CCAFS SLC-40': {
        'launch_site': [28.56319718, -80.57682003],
        'highway': [28.56324, -80.5708],
        'coastline': [28.56345, -80.568],
        'railroad': [28.56349, -80.58691],
        'city': [28.48921, -80.56274]
    },
    'KSC LC-39A': {
        'launch_site': [28.57325457, -80.64689529],
        'highway': [28.57325, -80.65548],
        'coastline': [28.58874, -80.61828],
        'railroad': [28.57318, -80.65396],
        'city': [28.48921, -80.56274]
    },
    'VAFB SLC-4E': {
        'launch_site': [34.63283416, -120.6107455],
        'highway': [34.64101, -120.45792],
        'coastline': [34.63365, -120.62593],
        'railroad': [34.63391, -120.62439],
        'city': [34.63893, -120.45784]
    }
}

# Add a marker and line for each site
for site, coords in sites.items():
    add_marker_and_line(site_map, site, 'Highway', *coords['launch_site'], *coords['highway'], 'blue')
    add_marker_and_line(site_map, site, 'Coastline', *coords['launch_site'], *coords['coastline'], 'green')
    add_marker_and_line(site_map, site, 'Railroad', *coords['launch_site'], *coords['railroad'], 'red')
    add_marker_and_line(site_map, site, 'City', *coords['launch_site'], *coords['city'], 'purple')



In [19]:
site_map

In [20]:
import pandas as pd

# Define the coordinates for each site
sites = {
    'CCAFS LC-40': {
        'launch_site': [28.56230197, -80.57735648],
        'highway': [28.56238, -80.57065],
        'coastline': [28.56253, -80.56786],
        'railroad': [28.56262, -80.5871],
        'city': [28.48921, -80.56274]
    },
    'CCAFS SLC-40': {
        'launch_site': [28.56319718, -80.57682003],
        'highway': [28.56324, -80.5708],
        'coastline': [28.56345, -80.568],
        'railroad': [28.56349, -80.58691],
        'city': [28.48921, -80.56274]
    },
    'KSC LC-39A': {
        'launch_site': [28.57325457, -80.64689529],
        'highway': [28.57325, -80.65548],
        'coastline': [28.58874, -80.61828],
        'railroad': [28.57318, -80.65396],
        'city': [28.48921, -80.56274]
    },
    'VAFB SLC-4E': {
        'launch_site': [34.63283416, -120.6107455],
        'highway': [34.64101, -120.45792],
        'coastline': [34.63365, -120.62593],
        'railroad': [34.63391, -120.62439],
        'city': [34.63893, -120.45784]
    }
}

# Initialize an empty list to store the data
data = []

# Loop through each site and calculate the distances
for site, coords in sites.items():
    highway_distance = calculate_distance(*coords['launch_site'], *coords['highway'])
    coastline_distance = calculate_distance(*coords['launch_site'], *coords['coastline'])
    railroad_distance = calculate_distance(*coords['launch_site'], *coords['railroad'])
    city_distance = calculate_distance(*coords['launch_site'], *coords['city'])
    
    # Append the data to the list
    data.append([site, coastline_distance, highway_distance, railroad_distance, city_distance])

# Create a DataFrame from the data
df = pd.DataFrame(data, columns=['Launch Site', 'Coastline Distance in km', 'Highway Distance in km', 'Railroad Distance in km', 'City Distance in km'])

# Display the DataFrame
df


Launch Site  Coastline Distance in km  Highway Distance in km  \
0   CCAFS LC-40                  0.928084                0.655233   
1  CCAFS SLC-40                  0.862105                0.588128   
2    KSC LC-39A                  3.283125                0.838578   
3   VAFB SLC-4E                  1.392655               14.015603   

   Railroad Distance in km  City Distance in km  
0                 0.952528             8.254539  
1                 0.986247             8.343822  
2                 0.690150            12.450739  
3                 1.254472            14.009983

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


##### Findings:

Are launch sites in close proximity to railways? 
Yes, all launch sites are in close proximity to railways. The distances range from approximately 0.69 km to 1.25 km, which suggests that railways are conveniently accessible for transportation of heavy launch equipment and materials.

Are launch sites in close proximity to highways? 
Yes, all launch sites except for VAFB SLC-4E are in close proximity to highways, with distances less than 1 km. For VAFB SLC-4E, the distance is approximately 14 km, which is relatively farther compared to the other sites but still within a reasonable distance.

Are launch sites in close proximity to the coastline? 
Yes, all launch sites are in close proximity to the coastline, with distances ranging from approximately 0.86 km to 3.28 km. This is likely because launching towards the sea provides a clear and safe trajectory for the rockets.

Do launch sites keep a certain distance away from cities? 
Yes, all launch sites maintain a considerable distance from cities, ranging from approximately 8.25 km to 14 km. This is likely for safety reasons, to minimize risks to populated areas in the event of a launch failure.

In summary, the location of launch sites appears to be strategically chosen to be near essential transportation routes (railways and highways), close to the coastline for safe launch trajectories, and at a safe distance from populated areas. These factors are likely considered to optimize logistics, safety, and operational efficiency. 

# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork865-2022-01-01)


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
